### PR TITLE
Eventable tracking

### PR DESCRIPTION
**Pull request by coderberry**
#82

### CODE REVIEW

**Comment by coderberry** [comment]
@hopsoft I am getting this error when running `./bin/standardize`

```[eberry:~/Code/code_fund_ads]$ ./bin/standardize                                                                                          (berry/82-event-log-tracking✱)
bundler: failed to load command: standardrb (/Users/eberry/.asdf/installs/ruby/2.5.3/bin/standardrb)
RuboCop::Runner::InfiniteCorrectionLoop: Infinite loop detected in /Users/eberry/Code/code_fund_ads/app/models/event.rb.
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:259:in `check_for_infinite_loop'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:242:in `block in iterate_until_no_changes'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:241:in `loop'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:241:in `iterate_until_no_changes'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:212:in `do_inspection_loop'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:115:in `block in file_offenses'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:133:in `file_offense_cache'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:113:in `file_offenses'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:486:in `call_with_index'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:455:in `process_incoming_jobs'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:437:in `block in worker'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:428:in `fork'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:428:in `worker'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:419:in `block in create_workers'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:418:in `each'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:418:in `each_with_index'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:418:in `create_workers'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:358:in `work_in_processes'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:264:in `map'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:217:in `each'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:53:in `warm_cache'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/rubocop-0.60.0/lib/rubocop/runner.rb:38:in `run'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/standard-0.0.14/lib/standard/runners/rubocop.rb:12:in `call'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/standard-0.0.14/lib/standard/cli.rb:18:in `run'
  /Users/eberry/.asdf/installs/ruby/2.5.3/lib/ruby/gems/2.5.0/gems/standard-0.0.14/exe/standardrb:7:in `<top (required)>'
  /Users/eberry/.asdf/installs/ruby/2.5.3/bin/standardrb:23:in `load'
  /Users/eberry/.asdf/installs/ruby/2.5.3/bin/standardrb:23:in `<top (required)>'```

**Comment by hopsoft** [comment]
You can get around this by removing any empty `protected` and `private` sections in the offending classes.

https://github.com/testdouble/standard/issues/28
https://github.com/rubocop-hq/rubocop/issues/6520

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  def set_property
+    @eventable = Property.find(params[:property_id])
+  end
+
+  def set_creative
+    @eventable = Creative.find(params[:creative_id])
+  end
+
+  def set_eventable
+    @eventable = GlobalID::Locator.locate_signed(params[:sgid])
```
Comment (hopsoft) [inline app/controllers/events_controller.rb]: Fancy 🎩

File: app/models/concerns/eventable.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    after_commit :track_creation, on: [:create]
+  end
+
+  module ClassMethods
+    # Helper method to lookup for events for a given object.
+    # This method is equivalent to obj.events
+    def find_events_for(obj)
+      Event.where(eventable_id: obj.id,
+                  eventable_type: obj.class.base_class.name).
+        order("created_at DESC")
```
Comment (hopsoft) [inline app/models/concerns/eventable.rb]: `order(created_at: :desc)`

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  # public instance methods ...................................................
+
+  # protected instance methods ................................................
+  protected
+
+  # private instance methods ..................................................
+  private
+
+end
```
Comment (hopsoft) [inline app/models/event.rb]: Needs annotations. `annotate`

File: app/models/event.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  #
+  #   irb>User.events.with_tags("registration", "email")
+
+  # additional config (i.e. accepts_nested_attribute_for etc...) ..............
+  tag_columns :tags
+
+  # class methods .............................................................
+  class << self
+    alias by_user find_events_by_user
+    alias for_eventable find_events_for_eventable
```
Comment (hopsoft) [inline app/models/event.rb]: Do we need aliases here or can this simply be the scope name?

Comment (coderberry) [reply app/models/event.rb]: good point

**Pull request merged**